---
title: "Ampliseq Results Report"
author: "Diagnostic Samples"
date: "`r Sys.Date()`"
output: html_document
params:
  csv: ""
  krona: ""
---

```{r setup, include=FALSE}
library(knitr)
library(ggplot2)
library(htmltools)
```


<br>
```{css, echo=FALSE}
.table caption {
    color: darkblue;
    font-weight: bold;
}
```

### Click dropdown to select sample {.tabset .tabset-dropdown}
```{r, results='asis',echo=FALSE}

samplelist <- read.csv(params$csv,header=FALSE,sep = ',')[-1,]
sampleid <- samplelist[,c(1)]

for (i in sampleid){
  cat("####", i, "\n")
  #use mappedreads.txt and display column 1 and 3
  mapped_reads <- (paste(i,"_mappedreads.txt",sep=""))
  df <- read.csv(mapped_reads,header = TRUE,sep=" ")
  mapped_stat<-df[,c(1,3)]
  colnames (mapped_stat) <- c("AMPLICON","MAPPED_READS")
  print(knitr::kable(mapped_stat,align="ll",caption = "NO. OF MAPPED READS. Generated using minimap2 and samtools"))
  cat('\n\n<!-- -->\n\n')
  fasta_id <- (paste(i,"_ORF.fasta",sep=""))
  fasta_file <- read.csv(fasta_id,header = FALSE )
  # Display the content
  cat("```\n")
  cat(fasta_file, sep = "\n")
  cat("\n```\n")
  # Encode the file content as base64
  encoded_content <- base64encode(charToRaw(paste(fasta_file, collapse = "\n")))
  # Get file name from file path and remove extra underscores
  file_name <- gsub("_+", "_", basename(fasta_id))
  # Provide a download link for the encoded FASTA content with file name
  cat('<a href="data:text/plain;charset=utf-8;base64,', encoded_content, '" download="', file_name, '">Download FASTA file</a>')
  
  cat('\n\n<!-- -->\n\n')
  }

```
<br>

<br>

## Taxonomic classification of raw reads 

```{r, echo=FALSE}
library("htmltools")
htmltools::tags$iframe(
  src = base64enc::dataURI(file= params$krona, mime="text/html; charset=UTF-8"),
  style="border:10; position:relative; top:0; left:; right::; bottom:; width:100%; height:800px"
)
```